pkg/scheduler/plugins/topology/topology_plugin_job_filtering.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2025 NVIDIA CORPORATION
+// SPDX-License-Identifier: Apache-2.0
+
+package topology
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/node_info"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/pod_info"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/podgroup_info"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/api/resource_info"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/log"
+	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/plugins/scores"
+	"k8s.io/apimachinery/pkg/types"
+	k8sframework "k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+const (
+	topologyAnnotationKey                   = "kai.scheduler/topology"
+	topologyRequiredPlacementAnnotationKey  = "kai.scheduler/topology-required-placement"
+	topologyPreferredPlacementAnnotationKey = "kai.scheduler/topology-preferred-placement"
+)
+
+type topologyStateData struct {
+	relevantDomains []*TopologyDomainInfo
+}
+
+func (t *topologyStateData) Clone() k8sframework.StateData {
+	return &topologyStateData{
+		relevantDomains: t.relevantDomains,
+	}
+}
+
+func (t *topologyPlugin) prePredicateFn(_ *pod_info.PodInfo, job *podgroup_info.PodGroupInfo) error {
+	topologyTree, err := t.getJobTopology(job)
+	if err != nil {
+		return err
+	}
+	if topologyTree == nil {
+		return nil
+	}
+
+	//Check cycle cache to see if the calculation has already been done
+	jobAllocateableDomains, _ := t.loadAllocateableDomainsFromCache(types.UID(job.PodGroupUID))
+	if jobAllocateableDomains != nil {
+		return nil
+	}
+
+	// Calc tree job allocation data
+	t.calcTreeAlocationData(job, topologyTree)
+
+	// Get best domains for the job
+	jobAllocateableDomain, err := t.getBestjobAllocateableDomains(job, topologyTree)
+	if err != nil {
+		return err
+	}
```

Comment:
Is there any case where these two are not called together? If not, just call `calcTreeAlocationData` from inside `getBestjobAllocateableDomains`

Reply:
I think that it makes sense to separate them as 2 different steps in the proccess:
1- Get matching topology
2- See if we have the result in cache
3- Calculate allocation data for the given job on all tree nodes
4- Get the domains which satisfies all the requirments
5- Save to cache